src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -175,6 +175,15 @@ private void SetGlobalErrorIfTrue(bool arg)
 
                     methodCompiler.CompileSynthesizedMethods(privateImplClass, diagnostics);
                 }
+
+                var rootModuleType = moduleBeingBuiltOpt.RootModuleType;
+                if (rootModuleType != null)
+                {
+                    // all threads that were adding methods must be finished now, we can freeze the class:
+                    rootModuleType.Freeze();
+
+                    methodCompiler.CompileSynthesizedMethods(rootModuleType, diagnostics);
+                }
```

Comment:
Direct copy of the lines above. In general, everything I did with RootModuleType is mimicking code that works with PrivateImplementationDetails.

1. Do you want this kind of generality?
2. Should any unification be done to avoid repetition of patterns between PrivateImplementationDetails and RootModuleType?

Reply:
seems to be only 2 right now, I'd wait for 3 before coming up with some abstraction :)

Reply:
> Do you want this kind of generality?

I think that for an injection of a single method with very simple signature, generalization or unification is not worth it. 

---
In reply to: [407010880](https://github.com/dotnet/roslyn/pull/43281#discussion_r407010880) [](ancestors = 407010880)

Reply:
I also do not think we would want to add more methods to the ```<Module>``` type in the near future, given limitations around them.

---
In reply to: [409707957](https://github.com/dotnet/roslyn/pull/43281#discussion_r409707957) [](ancestors = 409707957,407010880)